### DISCUSSION TITLE
More Code Blocks for Powered-Up/Mindstorms Color Sensor?

### DISCUSSION THREAD

**Discussion by LCR-Brick-Tech** [pybricks/support#1720] [Polls]
Are you interested in having more Pybricks code blocks for the Powered-Up/Mindstorms Robot Inventor Color Sensor?

This would be useful for creating more complex robot behaviors, especially for tasks like line following or color sorting where precise color detection is crucial.

One specific functionality I'd love to see is a block that allows us to turn off the built-in light on the color sensor while it detects colors in a loop. This is particularly important for projects like mine, where the sensor struggles to differentiate between black and white due to the brightness of the built-in light.

If you'd like to see an example of how these blocks could be used in a Lego Mindstorms vending machine project, check out the  [LCR Brick Tech Channel.](https://youtube.com/@lcrbricktech?si=yN4fJbOzJw6PCEOR)  Consider subscribing for a notification when the vending machine video is released!

**Comment by laurensvalk** [comment]
> I'd love to see is a block that allows us to turn off the built-in light on the color sensor

You can do that as follows:

![image](https://github.com/user-attachments/assets/7d9b3a32-a9f1-445f-aca4-2fecc3660b75)

However....

> the sensor struggles to differentiate between black and white

This isn't due to the light of the sensor. The sensor can tell the color of an object by emitting some light and measuring what comes back. So the sensor light will always turn back on when scanning the color of an object.

Instead, you'll want to _calibrate_ the sensor. You can tell it what the colors in your application look like, so the sensor will be able to differentiate them.

Here is a [tutorial](https://www.youtube.com/watch?v=kwnSbnlhOkA). This shows an example for color markers on train tracks, but it works just the same for candy in a vending machine using the Mindstorms Color Sensor.

**Reply by LCR-Brick-Tech** [reply]
Hi Laurens, thanks for the clarification about the sensor light.  That makes perfect sense!

I really appreciate the tip about sensor calibration. I'll definitely be exploring the tutorial you linked.

In the meantime, I had an idea for my vending machine project.  What if I placed a blue background behind the sensor? This way, when there's no object present, it would consistently read blue. Then, ideally, black objects would reflect minimal light and read as "zero," while white objects would still register as white.

Do you think this approach could be effective for differentiating black and white objects, or would calibration still be necessary?

This is alongside exploring calibration techniques, of course.

Thanks again for your help!

Additionally, I was wondering if there's a way to set a threshold for how precisely the sensor detects colors. For instance, could I define an HSV range, like H180 ± 15, to account for slight variations in the sensor reading?  This could be used alongside calibration for even more precise color detection. Thank you, Lorenzo.

**Comment by laurensvalk** [comment]
> I'll definitely be exploring the tutorial you linked. (...) In the meantime, I had an idea for my vending machine project. What if I placed a blue background behind the sensor? 

This is also covered in the tutorial :) In that case, the train tracks and the floor are the background. In your case, it could be whatever it sees behind the candy pieces you're scanning.

> Additionally, I was wondering if there's a way to set a threshold for how precisely the sensor detects colors. For instance, could I define an HSV range, like H180 ± 15, to account for slight variations in the sensor reading? This could be used alongside calibration for even more precise color detection.

This is already happening. It will round the measurement to the nearest matching color that you've set up. Depending on the number of colors you set up, it will work far beyond ± 15 degrees. This is what makes it so powerful compared to the LEGO method.

If you specify that your application has only red, blue, and black, for example, then "yellow" and "orange" will automatically rounded to red, rather than reporting incorrect values as you would get with the stock firmware.
